File: 11-mixing-milk/milk.cpp
/*
ID: caleb.h1
LANG: C++
PROG: milk
*/

#include <iostream>
#include <fstream>
#include <map>

#define MIN(a, b) a < b ? a : b

int main() {
    std::ifstream infile("milk.in");
    unsigned int milkToBuy;
    infile >> milkToBuy;
    unsigned int moneySpent = 0;

    unsigned int numOfFarmers;
    infile >> numOfFarmers;

    // map setup
    // prices<price, quantity>
    std::map<unsigned int, unsigned int> prices;

    unsigned int price;
    unsigned int quantity;
    for (size_t i = 0; i < numOfFarmers; ++i) {
        infile >> price;
        infile >> quantity;
        // If this is a new price
        if (prices.find(price) == prices.end())
            prices[price] = quantity;
        else
            prices[price] += quantity;
    }

    infile.close();


    // Go through each farmer (starting least expensive to move)
    // Either buy them out or buy as much as we need
    // No need to worry about the iterator b/c we are guarenteed enough milk is offered
    for (auto it = prices.begin(); milkToBuy > 0; ++it) {
        unsigned int order = MIN(milkToBuy, it->second);
        // it->second -= order; // Don't need to change quanitity since we'll either buy it all or be done 
        milkToBuy -= order;
        moneySpent += it->first * order;
    }

    std::ofstream outfile("milk.out");
    outfile << moneySpent << std::endl;
    outfile.close();


    return 0;
}

File: 11-mixing-milk/milk.py
"""
ID: caleb.h1
LANG: PYTHON3
PROG: milk
"""

def main():

    with open("milk.in", "r") as f:
        milkToBuy, numOfFarmers = (int(i) for i in f.readline().split())

        # dict setup
        # prices[price] = quantity
        prices = {}

        for _ in range(numOfFarmers):
            price, quantity = (int(i) for i in f.readline().split())
            # If this is a new price
            if not price in prices:
                prices[price] = quantity
            else:
                prices[price] += quantity

    keys = sorted(prices)

    moneySpent = 0

    # Go through each farmer (starting least expensive to move)
    # Either buy them out or buy as much as we need
    # No need to worry about the iterator b/c we are guarenteed enough milk is offered
    for price in keys:
        if milkToBuy <= 0:
            break
        order = min(milkToBuy, prices[price])
        milkToBuy -= order
        moneySpent += price * order
    
    with open("milk.out", "w") as f:
        f.write(f'{moneySpent}\n')

if __name__ == '__main__':
    main()